### PR TITLE
Design paterns

### PR DESCRIPTION
**Pull request by luisirio**
This PR is related to assignment 4.

### CODE REVIEW

File: life_expectancy/loading_saving.py [hunk 1]
```diff
@@ -1,29 +1,51 @@
...
 
-    Returns:
-        [pd.DataFrame]: Dataframe to be cleaned.
-    """
-    data_df = pd.read_csv(file_path, sep='\t')
+    def load_json(self) -> pd.DataFrame:
+        """Method to load json file"""
+        return pd.read_json(JSON_FILE_PATH)
+
+    def load_data(self) -> None:
```
Comment (MeloDuarte) [inline life_expectancy/loading_saving.py:33]: In my opinion the code would be cleaner if you removed the methods `load_csv` and `load_json`:
```python
def load_data(self) -> None:
        """Method to verify data format"""
        if self.data_format == 'csv':
            return pd.read_csv(TSV_FILE_PATH, sep='\t')
        if self.data_format == 'json':
            return pd.read_json(JSON_FILE_PATH)
        raise ValueError("Unsupported data format.")
```

File: life_expectancy/tests/test_main.py [hunk 2]
```diff
@@ -7,12 +7,26 @@
 from life_expectancy.main import main
 
 
-def test_main(pt_life_expectancy_expected):
-    """Run the `main` function and compare the output to the expected output"""
+def test_main_csv(pt_life_expectancy_expected):
```
Comment (MeloDuarte) [inline life_expectancy/tests/test_main.py:10]: The behavior of the main function should not be agnostic to file types? assuming that you have also tested the cleaning function for both file types...

**Review by MeloDuarte — CHANGES_REQUESTED** [review]

File: life_expectancy/cleaning.py [hunk 3]
```diff
@@ -1,46 +1,98 @@
...
+            cleaned_df = self._clean_csv(data_df, country)
+        elif self.data_format == 'json':
+            cleaned_df = self._clean_json(data_df, country)
+        else:
+            raise ValueError("Unsupported data format.")
+
+        return cleaned_df
+
+    @staticmethod
+    def _clean_csv(data_df: pd.DataFrame, country: Region):
```
[the remark this answers is not in the document]
Comment (luisirio) [reply life_expectancy/cleaning.py]: Resolved

File: life_expectancy/loading_saving.py (hunk 1 above)
Comment (luisirio) [reply life_expectancy/loading_saving.py:25]: Resolved

File: life_expectancy/region.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    SI = "SI"
+    SK = "SK"
+    SM = "SM"
+    TR = "TR"
+    UA = "UA"
+    UK = "UK"
+    XK = "XK"
+
+    @classmethod
+    def list_countries(cls) -> List[str]:
```
[the remark this answers is not in the document]
Comment (luisirio) [reply life_expectancy/region.py]: Resolved

File: life_expectancy/tests/test_main.py (hunk 2 above)
Comment (luisirio) [reply life_expectancy/tests/test_main.py:10]: Resolved

**Review by MeloDuarte — APPROVED** [review]
LGTM 🍤

File: life_expectancy/main.py [hunk 5]
```diff
@@ -3,31 +3,36 @@
...
-from life_expectancy.loading_saving import load_data, save_data
+from life_expectancy.cleaning import DataCleaner
+from life_expectancy.loading_saving import DataLoader, save_data
 
 PARENT_PATH = Path(__file__).parent
 FILE_PATH = PARENT_PATH / 'data'
-INPUT_FILE_NAME = 'eu_life_expectancy_raw.tsv'
 OUTPUT_FILE_NAME = 'pt_life_expectancy.csv'
-INPUT_FILE_PATH = FILE_PATH / INPUT_FILE_NAME
 OUTPUT_FILE_PATH = FILE_PATH / OUTPUT_FILE_NAME
```
Comment (MrCordeiro) [inline life_expectancy/main.py:12]: These global contants are also being define on loading_saving. If, in the future you need to change them, but forget to change both, you will have errors. Try to remove this duplication by either passing the value as parameters or making these modules import the same constant

File: life_expectancy/main.py [hunk 6]
```diff
@@ -3,31 +3,36 @@
...
+
 
 if __name__ == "__main__":
     parser = argparse.ArgumentParser()
+    parser.add_argument('file_path')
     parser.add_argument('region')
     args = parser.parse_args()
-    main(args.region)
\ No newline at end of file
+    main(args.file_path, args.region)
```
Comment (MrCordeiro) [inline life_expectancy/main.py:38]: Your main is asking for a file format, but you're passing a file path.

File: life_expectancy/loading_saving.py [hunk 7]
```diff
@@ -1,29 +1,43 @@
...
-    Returns:
-        [pd.DataFrame]: Dataframe to be cleaned.
+class DataLoader:
+    """
+    This class is responsible for verifying the format
+    and assign the right load method
     """
-    data_df = pd.read_csv(file_path, sep='\t')
+    def __init__(self, data_format):
+        self.data_format = data_format
```
Comment (MrCordeiro) [inline life_expectancy/loading_saving.py:23]: **Suggestion**: make your class receive a Path object and infer the format from there

```suggestion
    def __init__(self, file_path: pathlib.Path) -> None:
        self.file_path = file_path
        self.data_format = file_path.suffix
```

File: life_expectancy/loading_saving.py [hunk 8]
```diff
@@ -1,29 +1,43 @@
...
+
+    def load_data(self) -> None:
+        """Method to verify data format"""
+        if self.data_format == 'csv':
+            return pd.read_csv(TSV_FILE_PATH, sep='\t')  # load csv file
+        if self.data_format == 'json':
+            return pd.read_json(JSON_FILE_PATH)  # load json file
```
Comment (MrCordeiro) [inline life_expectancy/loading_saving.py:30]: If you have a file path attribute, you can get rid of the duplicate constants:

```suggestion
        if self.data_format == '.csv':
            return pd.read_csv(self.file_path, sep='\t')
        if self.data_format == '.json':
            return pd.read_json(self.file_path)
```

File: life_expectancy/main.py [hunk 9]
```diff
@@ -3,31 +3,36 @@
...
     """
-    data_df = load_data(INPUT_FILE_PATH)
-    clean_df = clean_data(data_df, region)
-    save_data(clean_df, OUTPUT_FILE_PATH)
-    return clean_df
+
+    data_df = DataLoader(file_format).load_data()
+
+    cleaned_df = DataCleaner(file_format).clean_data(data_df, region)
+
```
Comment (MrCordeiro) [inline life_expectancy/main.py:28]: **Suggestion**: One problem I can see here is that you have to define the logic to select the correct treatment in 2 places (inside DataLoader and DataCleaner). If you add another compenent to this pipeline, you will need to add logic inside a 3rd object, and so forth. You are really not taking advantage of the Strategy Pattern as the number of places you need to change increase with the number of components.

My suggestion is for you to rearrange your code so that the logic to define the file type is done inside `main`. Yes, I know this would invalidade some of my previous suggestions... 😬

But it would look something like this:

```suggestion
    if file_path.suffix == ".tsv":
        loader, cleaner = DataLoaderTSV(), DataCleanerTSV()
    elif file_path.suffix == ".json":
        loader, cleaner = DataLoaderJSON(), DataCleanerJSON()

    df = loader.load_data()
    cleaned_dv = cleaner.clean_data(df, region)
```

It's like lego. You submodule just hold the your lego pieces (classes and functions) and your main.py is where the logic is applied and the pieces are put together. This is works for small repos such as these and it's very easy to maintain.

**Review by MrCordeiro — COMMENTED** [review]
Left a few comments!

File: life_expectancy/main.py (hunk 5 above)
Comment (luisirio) [reply life_expectancy/main.py:13]: Thank you for your comment @MrCordeiro! I just changed this after your comment in PDS. 
Which variable is duplicated?

File: life_expectancy/loading_saving.py (hunk 7 above)
Comment (luisirio) [reply life_expectancy/loading_saving.py]: Got it! Thank you!

File: life_expectancy/loading_saving.py (hunk 8 above)
Comment (luisirio) [reply life_expectancy/loading_saving.py]: Got it! Thank you!

File: life_expectancy/main.py (hunk 9 above)
Comment (luisirio) [reply life_expectancy/main.py]: I changed the code according your suggestions! Thank you!

File: life_expectancy/tests/test_loading_saving.py [hunk 10]
```diff
@@ -5,22 +5,28 @@
 
 import pandas as pd
 
-from life_expectancy.loading_saving import load_data, save_data
+from life_expectancy.loading_saving import DataLoader, save_data
 
 PARENT_PATH = Path(__file__).parent.parent
 FILE_PATH = PARENT_PATH / 'data'
```
[the remark this answers is not in the document]
Comment (luisirio) [reply life_expectancy/tests/test_loading_saving.py:12]: FIXTURES_DIR is used in conftest.py

**Comment by luisirio** [comment]
@MrCordeiro I have implemented your suggestions in this [commit](e9fbf4ce56711cc15e86344b69516526da67902e)!
Thank you for your valuable review!

File: life_expectancy/tests/test_loading_saving.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (MrCordeiro) [reply life_expectancy/tests/test_loading_saving.py:12]: You are building `INPUT_FILE_PATH` and `OUTPUT_FILE_PATH` using the `FILE_PATH` variable. However, this means that your tests will touch the `data/` directory, which is something we don't want.

Instead, you should make your tests use the FIXTURE_DIR. 

Take your `test_load_data_tsv`, for example. You are loading your input data from `data/` and the expected from `fixtures/`, but both of them should be being read from `/fixtures`.

File: life_expectancy/main.py (hunk 5 above)
Comment (MrCordeiro) [reply life_expectancy/main.py:13]: Oh, you've since removed them! Great!
